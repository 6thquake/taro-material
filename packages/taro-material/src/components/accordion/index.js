import Taro from '@tarojs/taro'
import PropTypes from 'prop-types'
import classNames from 'classnames'
import { View, Text } from '@tarojs/components'
import AtComponent from '../../common/component'
import { delayQuerySelector, initTestEnv } from '../../common/utils'

initTestEnv()

// 文档
export default class AtAccordion extends AtComponent {
  constructor () {
    super(...arguments)
    this.isCompleted = true
    this.startOpen = false
    this.state = {
      wrapperHeight: ''
    }
  }

  handleClick = event => {
    const { open } = this.props
    if (!this.isCompleted) return

    this.props.onClick(!open, event)
  }

  toggleWithAnimation () {
    const { open, isAnimation } = this.props
    if (!this.isCompleted || !isAnimation) return

    this.isCompleted = false
    delayQuerySelector(this, '.at-accordion__body', 0)
      .then(rect => {
        const height = parseInt(rect[0].height)
        const startHeight = open ? height : 0
        const endHeight = open ? 0 : height
        this.startOpen = false
        this.setState({
          wrapperHeight: startHeight
        }, () => {
          setTimeout(() => {
            this.setState({
              wrapperHeight: endHeight
            }, () => {
              setTimeout(() => {
                this.isCompleted = true
                this.setState({})
              }, 700)
            })
          }, 100)
        })
      })
  }

  componentWillReceiveProps (nextProps) {
    if (nextProps.open !== this.props.open) {
      this.startOpen = nextProps.open && nextProps.isAnimation
      this.toggleWithAnimation()
    }
  }

  render () {
    const {
      customStyle,
      className,
      title,
      icon,
      hasBorder,
      open
    } = this.props
    const { wrapperHeight } = this.state

    const rootCls = classNames('at-accordion', className)
    const iconCls = classNames({
      'at-icon': true,
      [`at-icon-${icon && icon.value}`]: icon && icon.value,
      'at-accordion__icon': true,
    })
    const headerCls = classNames('at-accordion__header', {
      'at-accordion__header--noborder': !hasBorder
    })
    const arrowCls = classNames('at-accordion__arrow', {
      'at-accordion__arrow--folded': !!open
    })
    const contentCls = classNames('at-accordion__content', {
      'at-accordion__content--inactive': (!open && this.isCompleted) || this.startOpen
    })
    const iconStyle = {
      color: (icon && icon.color) || '',
      fontSize: (icon && `${icon.size}px`) || '',
    }
    const contentStyle = { height: `${wrapperHeight}px` }

    if (this.isCompleted) {
      contentStyle.height = ''
    }

    return <View className={rootCls} style={customStyle}>
      <View className={headerCls} onClick={this.handleClick}>
        {icon && icon.value && <Text className={iconCls} style={iconStyle}></Text>}
        <View className='at-accordion__title'>{title}</View>
        <View className={arrowCls}>
          <Text className='at-icon at-icon-chevron-down'></Text>
        </View>
      </View>
      <View style={contentStyle} className={contentCls}>
        <View className='at-accordion__body'>
          {this.props.children}
        </View>

      </View>
    </View>
  }
}

AtAccordion.defaultProps = {
  open: false,
  customStyle: '',
  className: '',
  title: '',
  icon: {},
  hasBorder: true,
  isAnimation: true,
  onClick: () => {},
}

AtAccordion.propTypes = {
  customStyle: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.string
  ]),
  className: PropTypes.oneOfType([
    PropTypes.array,
    PropTypes.string
  ]),
  open: PropTypes.bool,
  isAnimation: PropTypes.bool,
  title: PropTypes.string,
  icon: PropTypes.object,
  hasBorder: PropTypes.bool,
  onClick: PropTypes.func,
}
