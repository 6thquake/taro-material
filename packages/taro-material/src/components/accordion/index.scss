@import '../../style/theme/default.scss';
@import '../../style/mixins/index.scss';

.at-accordion {
  &__header {
    position: relative;
    display: flex;
    box-sizing: border-box;
    width: 100%;
    min-height: $font-size-lg;
    font-size: $font-size-lg;
    padding: $spacing-v-lg $spacing-h-lg;
    @include hairline-bottom();
    @include active;
  }

  &__title {
    flex: 1;
    text-align: left;
    @include line;
  }

  &__content {
    position: relative;
    width: 100%;
    transition: height 0.3s;
    overflow: hidden;
    @include hairline-top();
  }
}
