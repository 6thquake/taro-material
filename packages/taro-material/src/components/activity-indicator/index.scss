@import "../../style/mixins/index.scss";
@import "../../style/theme/default.scss";

.at-activity-indicator {
  @include display-flex();

  line-height: $line-height-base;

  &--center {
    @include absolute-center;
  }

  &__body {
    @include flex(0, 0, auto);

    line-height: 0;
  }

  &__content {
    @include flex(0, 0, auto);
    @include align-self(center);

    font-size: 28px;
    margin-left: $spacing-h-lg;
    color: $color-grey-2;
  }
}
