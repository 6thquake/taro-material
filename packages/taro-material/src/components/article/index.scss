@import '../../style/theme/default.scss';

.at-article {
  word-break: break-all;
  word-wrap: break-word;
  line-height: $line-height-zh;

  &__h1,
  &__h2,
  &__h3 {
    color: $color-grey-0;
    margin: 0 30px;
  }

  &__h1 {
    margin-top: 30px;
    font-size: 48px;
    line-height: 72px;
  }

  &__info {
    font-size: $font-size-base;
    color: $color-grey-3;
    line-height: 42px;
    margin: 0 30px;
  }

  &__h2 {
    font-size: $font-size-xxl;
    line-height: 60px;
    margin-top: 30px;
  }

  &__h3 {
    font-size: $font-size-lg;
    line-height: 48px;
  }

  &__p {
    font-size: $font-size-base;
    color: $color-grey-1;
    line-height: 42px;
    margin: 25px 30px 0;
  }

  &__img {
    width: 690px;
    display: block;
    margin: 20px auto 0;
    border-radius: 4px;
  }
}
