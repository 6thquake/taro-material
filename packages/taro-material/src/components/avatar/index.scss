@import '../../style/theme/default.scss';

$avatar-size-sm: 80px;
$avatar-size-md: 100px;
$avatar-size-lg: 120px;

.at-avatar {
  width: $avatar-size-md;
  height: $avatar-size-md;
  color: #fff;
  font-size: $font-size-xxl;
  line-height: $avatar-size-md;
  text-align: center;
  border-radius: $border-radius-md;
  background: $color-grey-4;
  box-shadow: 0 0 50px 0 rgba(0, 0, 0, 0.05);
  overflow: hidden;

  &__img {
    width: 100%;
    height: 100%;
  }

  &--large {
    width: $avatar-size-lg;
    height: $avatar-size-lg;
    font-size: 48px;
    line-height: $avatar-size-lg;
  }

  &--small {
    width: $avatar-size-sm;
    height: $avatar-size-sm;
    font-size: $font-size-lg;
    line-height: $avatar-size-sm;
  }

  &--circle {
    border-radius: 50%;
    background-clip: border-box;
    overflow: hidden;
  }
}
