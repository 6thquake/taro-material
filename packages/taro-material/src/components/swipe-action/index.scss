@import "../../style/mixins/index.scss";
@import "../../style/theme/default.scss";

.at-swipe-action {
  position: relative;
  overflow: hidden;

  &__content {
    background-color: white;
    position: relative;
    z-index: 2;
    font-size: $font-size-lg;

    &.animtion {
      transition: transform 300ms $timing-func;
    }
  }

  &__option {
    @include active;
    @include align-items(center);

    height: 100%;
    background-color: $color-grey-2;
    display: inline-flex;
    color: $color-white;
    font-size: $font-size-base;
    padding: 0 $spacing-h-xl;
    text-align: center;
  }
}
