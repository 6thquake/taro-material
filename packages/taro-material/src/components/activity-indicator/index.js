import Taro from '@tarojs/taro'
import { View, Text } from '@tarojs/components'

import PropTypes from 'prop-types'
import classNames from 'classnames'

import AtLoading from '../loading/index'
import AtComponent from '../../common/component'

import './index.scss'

export default class AtActivityIndicator extends AtComponent {
  render () {
    const { color, size, mode, content } = this.props

    const rootClass = classNames(
      'at-activity-indicator',
      {
        'at-activity-indicator--center': mode === 'center'
      },
      this.props.className
    )

    return (
      <View className={rootClass}>
        <View className='at-activity-indicator__body'>
          <AtLoading size={size} color={color} />
        </View>
        {content && (
          <Text className='at-activity-indicator__content'>{content}</Text>
        )}
      </View>
    )
  }
}

AtActivityIndicator.defaultProps = {
  size: 24,
  color: '#6190E8'
}

AtActivityIndicator.propTypes = {
  size: PropTypes.number,
  mode: PropTypes.string,
  color: PropTypes.string,
  content: PropTypes.string
}
