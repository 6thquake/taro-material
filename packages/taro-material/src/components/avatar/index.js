import Taro from '@tarojs/taro'
import { View, Image, Text, OpenData } from '@tarojs/components'
import PropTypes from 'prop-types'
import classNames from 'classnames'

import AtComponent from '../../common/component'

import './index.scss'

const SIZE_CLASS = {
  large: 'large',
  normal: 'normal',
  small: 'small',
}

export default class AtAvatar extends AtComponent {
  constructor () {
    super(...arguments)
    this.state = {
      isWEAPP: Taro.getEnv() === Taro.ENV_TYPE.WEAPP,
    }
  }

  onClick () {
  }

  render () {
    const {
      size,
      circle,
      image,
      text,
      openData,
      customStyle,
    } = this.props
    const rootClassName = ['at-avatar']

    const classObject = {
      [`at-avatar--${SIZE_CLASS[size]}`]: SIZE_CLASS[size],
      'at-avatar--circle': circle,
    }

    let letter = ''
    if (text) letter = text[0]

    let elem
    if (openData && openData.type === 'userAvatarUrl' && this.state.isWEAPP) {
      elem = (<OpenData type={openData.type}></OpenData>)
    } else if (image) {
      elem = (<Image className='at-avatar__img' src={image} />)
    } else {
      elem = (<Text className='at-avatar__text'>{letter}</Text>)
    }
    return (
      <View
        className={classNames(rootClassName, classObject, this.props.className)}
        style={customStyle}
      >{elem}</View>
    )
  }
}

AtAvatar.defaultProps = {
  size: 'normal',
  circle: false,
  text: '',
  image: '',
  openData: {},
  customStyle: {},
  className: '',
}

AtAvatar.propTypes = {
  size: PropTypes.oneOf(['large', 'normal', 'small']),
  circle: PropTypes.bool,
  text: PropTypes.string,
  image: PropTypes.string,
  openData: PropTypes.object,
  customStyle: PropTypes.oneOfType([PropTypes.object, PropTypes.string]),
  className: PropTypes.oneOfType([PropTypes.array, PropTypes.string]),
}
