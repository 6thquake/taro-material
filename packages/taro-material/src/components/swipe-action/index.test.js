import Nerv, { findDOMNode } from 'nervjs'
import { renderToString } from 'nerv-server'
import { renderIntoDocument } from 'nerv-test-utils'
import simulant from 'simulant'

import { View } from '@tarojs/components'

import AtSwipeAction from '../../../.temp/components/swipe-action/index'

const MAX_OFFSET_SIZE = 101

const Simulate = {}
const EVENTS = [
  'keyDown',
  'keyPress',
  'keyUp',
  'focus',
  'blur',
  'click',
  'contextMenu',
  'doubleClick',
  'drag',
  'dragEnd',
  'dragEnter',
  'dragExit',
  'dragLeave',
  'dragOver',
  'dragStart',
  'drop',
  'mouseDown',
  'mouseEnter',
  'mouseLeave',
  'mouseMove',
  'mouseOut',
  'mouseOver',
  'mouseUp',
  'change',
  'input',
  'submit',
  'touchCancel',
  'touchEnd',
  'touchMove',
  'touchStart',
  'load',
  'error',
  'animationStart',
  'animationEnd',
  'animationIteration',
  'transitionEnd'
]
EVENTS.forEach(event => {
  Simulate[event] = (node, mock) =>
    simulant.fire(node, event.toLowerCase(), mock)
})

const OPTIONS = [
  {
    text: '取消',
    className: 'cancel',
    style: {
      backgroundColor: '#6190E8'
    }
  },
  {
    text: '确认',
    className: 'confirm',
    style: {
      backgroundColor: '#FF4949'
    }
  }
]

describe('SwipeAction Snap', () => {
  it('render options', () => {
    const component = renderToString(
      <AtSwipeAction
        className='swipe-action--test'
        disabled
        autoClose
        options={OPTIONS}
      >
        <View className='normal'>AtSwipeAction 一般使用场景</View>
      </AtSwipeAction>
    )
    expect(component).toMatchSnapshot()
  })
})

describe('SwipeAction Swipe Behavior', () => {
  const onClick = jest.fn()
  const onClosed = jest.fn()
  const onOpened = jest.fn()

  const component = renderIntoDocument(
    <AtSwipeAction
      options={OPTIONS}
      onClick={onClick}
      onClosed={onClosed}
      onOpened={onOpened}
    >
      <View className='normal'>AtSwipeAction 一般使用场景</View>
    </AtSwipeAction>
  )
  const componentDom = findDOMNode(component, 'at-swipe-action')
  component.maxOffsetSize = MAX_OFFSET_SIZE

  beforeEach(() => {
    onClick.mockReset()
    onClosed.mockReset()
    onOpened.mockReset()
    component._reset()
  })

  it('SwipeAction onClick', () => {
    const optionDom = componentDom.querySelector('.at-swipe-action__option')

    Simulate.click(optionDom)

    expect(onClick).toBeCalled()
  })

  it('SwipeAction Swipe Right', () => {
    // 一开始向右滑动是没有效果的
    expect(component.isTouching).toBeFalsy()
    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })
    expect(component.isTouching).toBeTruthy()

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: 100 }]
    })

    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(0)

    Simulate.touchEnd(componentDom)
    expect(onClosed).not.toBeCalled()
    expect(onOpened).not.toBeCalled()
    expect(component.state.isOpened).toBeFalsy()
    expect(component.state.offsetSize).toEqual(0)
  })

  it('SwipeAction Swipe Left', () => {
    // 一开始向左滑动 直到最大时将停止
    expect(component.isTouching).toBeFalsy()
    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })
    expect(component.isTouching).toBeTruthy()

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -110 }]
    })

    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(-110)

    Simulate.touchEnd(componentDom)
    component.forceUpdate()

    expect(onOpened).toBeCalled()
    expect(onClosed).not.toBeCalled()
    expect(component.state.isOpened).toBeTruthy()
    expect(component.state.offsetSize).toEqual(-MAX_OFFSET_SIZE)
  })

  it('SwipeAction Completed Swipe Action', () => {
    // 开始向左滑懂
    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -100 }]
    })

    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(-100)

    // 向左滑动结束
    Simulate.touchEnd(componentDom)
    component.forceUpdate()

    expect(onOpened).toBeCalled()
    expect(component.state.isOpened).toBeTruthy()
    expect(component.endValue).toEqual(-MAX_OFFSET_SIZE)
    expect(component.state.offsetSize).toEqual(-MAX_OFFSET_SIZE)

    // 开始向右滑动
    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: 100 }]
    })

    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(-1)

    // 向右滑动结束
    Simulate.touchEnd(componentDom)
    component.forceUpdate()
    expect(onClosed).toBeCalled()
    expect(component.endValue).toEqual(0)
    expect(component.state.isOpened).toBeFalsy()
    expect(component.state.offsetSize).toEqual(0)
  })
})

describe('SwipeAction Props', () => {
  it('SwipeAction Disabled', () => {
    const component = renderIntoDocument(
      <AtSwipeAction disabled options={OPTIONS}>
        <View className='normal'>AtSwipeAction 一般使用场景</View>
      </AtSwipeAction>
    )
    const componentDom = findDOMNode(component, 'at-swipe-action')
    component.maxOffsetSize = MAX_OFFSET_SIZE

    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -110 }]
    })
    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(0)
  })

  it('SwipeAction Disabled', () => {
    const component = renderIntoDocument(
      <AtSwipeAction disabled options={OPTIONS}>
        <View className='normal'>AtSwipeAction 一般使用场景</View>
      </AtSwipeAction>
    )
    const componentDom = findDOMNode(component, 'at-swipe-action')
    component.maxOffsetSize = MAX_OFFSET_SIZE

    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -110 }]
    })
    component.forceUpdate()
    expect(component.state.offsetSize).toEqual(0)
  })

  it('SwipeAction AutoClose', () => {
    const component = renderIntoDocument(
      <AtSwipeAction autoClose options={OPTIONS}>
        <View className='normal'>AtSwipeAction 一般使用场景</View>
      </AtSwipeAction>
    )
    const componentDom = findDOMNode(component, 'at-swipe-action')
    const swipeActionButtonDom = componentDom.querySelector(
      '.at-swipe-action__option'
    )
    component.maxOffsetSize = MAX_OFFSET_SIZE

    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })
    component.forceUpdate()

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -110 }]
    })
    component.forceUpdate()

    Simulate.touchEnd(componentDom)
    component.forceUpdate()

    expect(component.state.isOpened).toBeTruthy()
    expect(component.state.offsetSize).toEqual(-MAX_OFFSET_SIZE)

    Simulate.click(swipeActionButtonDom)
    component.forceUpdate()

    expect(component.endValue).toEqual(0)
    expect(component.isTouching).toBeFalsy()
    expect(component.state.isOpened).toBeFalsy()
    expect(component.state.offsetSize).toEqual(0)
  })

  it('SwipeAction IsClose', () => {
    const component = renderIntoDocument(
      <AtSwipeAction autoClose options={OPTIONS}>
        <View className='normal'>AtSwipeAction 一般使用场景</View>
      </AtSwipeAction>
    )
    const componentDom = findDOMNode(component, 'at-swipe-action')
    component.maxOffsetSize = MAX_OFFSET_SIZE

    Simulate.touchStart(componentDom, {
      touches: [{ clientX: 0 }]
    })
    component.forceUpdate()

    Simulate.touchMove(componentDom, {
      touches: [{ clientX: -110 }]
    })
    component.forceUpdate()

    Simulate.touchEnd(componentDom)
    component.forceUpdate()

    expect(component.state.isOpened).toBeTruthy()
    expect(component.state.offsetSize).toEqual(-MAX_OFFSET_SIZE)

    const spy = jest
      .spyOn(component, 'componentWillReceiveProps')
      .bind(component)
    spy({ isClose: true })

    component.forceUpdate()

    expect(component.state.isOpened).toBeFalsy()
  })
})
