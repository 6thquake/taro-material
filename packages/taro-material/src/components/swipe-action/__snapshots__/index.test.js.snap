// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SwipeAction Snap render options 1`] = `"<div class=\\"at-swipe-action swipe-action--test\\"><div style=\\"transform:translate3d(0px,0,0);\\" class=\\"at-swipe-action__content animtion\\"><div class=\\"normal\\">AtSwipeAction 一般使用场景</div></div><div class=\\"at-swipe-action__options\\"><div style=\\"background-color:#6190E8;\\" class=\\"at-swipe-action__option cancel\\"><span class=\\"taro-text option__text\\">取消</span></div><div style=\\"background-color:#FF4949;\\" class=\\"at-swipe-action__option confirm\\"><span class=\\"taro-text option__text\\">确认</span></div></div></div>"`;
